### ISSUE TITLE
Focus is not handled properly for dialogs.

### ISSUE THREAD

**Issue by Nancy-Salpepi** [phetsims/fourier-making-waves#194]
**Test device**
MacBook Air (m1 chip)

**Operating System**
11.6

**Browser**
chrome

**Problem description**
https://github.com/phetsims/qa/issues/711

When using keyboard nav, the close (X) button is already highlighted when the Keyboard Shortcuts dialog is opened. However, when the Info and Expand Sum dialogs are opened, it is not. This prevents the user from being able to hit "esc" to exit. In those instances, the "tab"  button must first be pressed and then either "esc" or "space bar" to exit. 

**Visuals**
<img width="493" alt="Screen Shot 2021-09-28 at 3 58 42 PM" src="https://user-images.githubusercontent.com/87318828/135158818-e33c0dba-e0aa-48b2-8970-620dd3522625.png">

<details>

Troubleshooting information:
!!!!! DO NOT EDIT !!!!!
Name: ‪Fourier: Making Waves‬
URL: https://phet-dev.colorado.edu/html/fourier-making-waves/1.0.0-rc.1/phet/fourier-making-waves_all_phet.html
Version: 1.0.0-rc.1 2021-09-28 15:44:23 UTC
Features missing: applicationcache, applicationcache, touch
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.61 Safari/537.36
Language: en-US
Window: 1391x690
Pixel Ratio: 2/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 31 uniform: 1024
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 80)
Max viewport: 16384x16384
OES_texture_float: true
Dependencies JSON: {}
</details>

**Comment by pixelzoom** [comment]
@jessegreenberg please provide guidance here.  Is there something that needs to be done to set focus when a Dialog is opened? Why doesn't Dialog handle that automatically?

**Comment by pixelzoom** [comment]
Here's the relevant code related to opening a `KeyboardHelpDialog`, in KeyboardHelpDialog.js:

```js
    keyboardHelpDialogCapsule = new PhetioCapsule( tandem => {
      const keyboardHelpDialog = new KeyboardHelpDialog( content, {
        tandem: tandem
      } );
      keyboardHelpDialog.setFocusOnCloseNode( this );
      return keyboardHelpDialog;
    }, [], {
      tandem: tandem.createTandem( 'keyboardHelpDialogCapsule' ),
      phetioType: PhetioCapsule.PhetioCapsuleIO( Dialog.DialogIO )
    } );

...

    options.listener = () => {
      const keyboardHelpDialog = keyboardHelpDialogCapsule.getElement();
      keyboardHelpDialog.show();

      // if listener was fired because of accessibility
      if ( this.isPDOMClicking() ) {

        // focus the close button if the dialog is open with a keyboard
        keyboardHelpDialog.focusCloseButton();
      }
    };
```

So is every client that creates an accessible Dialog expected to call `setFocusOnCloseNode` after instantiation, and `focusCloseButton` after `show`?  There's a lot of boilerplate that I'll have to copy to make this work.

**Comment by pixelzoom** [comment]
I've added support for this in the above commits, in master and 1.0 branches.

Here's the boilerplate that I had to add in 3 places to make this work:

```js
    const dialog = new SomeDialog(...);

    const button = new SomeButton( {
      listener: () => {
        dialog.show();
        if ( button.isPDOMClicking() ) {
          dialog.focusCloseButton();
        }
      },
      ...
    } );

    dialog.setFocusOnCloseNode( button );
```

@jessegreenberg is this the correct pattern?

**Comment by jessegreenberg** [comment]
The change you made will work and matches what is currently supported by Dialog. And that should fly for 1.0 release.

The pattern can generally be improved though and this can be handled in Dialog now. Checking the source of input with `isPDOMClicking` or `SceneryEvent.isFromPDOM` used to be required because we didn't want the highlight to appear from mouse or touch. But the highlight implementation has been improved so that this isn't necessary. And `setFocusOnCloseNode` shouldn't be needed, it is only needed if focus should return to something other than the Node that opened the Dialog.

The right way forward is to focus the close button by default when the Dialog is opened but offer an option to focus something else instead.

**Comment by pixelzoom** [comment]
Thanks @jessegreenberg. I removed the unnecessary calls to `setFocusOnCloseNode` in the above commits to master and 1.0.

Oh rats... I found another dialog that I need to handle, the Info dialog for Wave Game screen.

**Comment by pixelzoom** [comment]
Wave Game info dialog is handled in the above commits, for master and 1.0.

**Comment by pixelzoom** [comment]
Slack conversation with @jessegreenberg 


Chris Malley  3:13 PM
Why is it necessary to do this:
```js
        if ( infoButton.isPDOMClicking() ) {
          infoDialog.focusCloseButton();
        }
```
… instead of simply calling `focusCloseButton`.

Jesse Greenberg  3:16 PM
It was historically necessary but isn't anymore. I tried to describe in https://github.com/phetsims/fourier-making-waves/issues/194#issuecomment-929620651 - it used to be that the focus highlight would appear as soon as focus() was called so we wanted to prevent that for mouse and touch input. But that is not the case anymore so isPDOMClicking checks shouldn't be needed either.

Chris Malley  3:18 PM
OK thanks, I think I’ll remove them then. Trying to pare it down to only what’s necessary, which seems to be calling `dialog.focusCloseButton()`.

**Comment by Nancy-Salpepi** [comment]
@pixelzoom should the change be made for each amplitude box as well?

<img width="170" alt="Screen Shot 2021-09-28 at 5 21 43 PM" src="https://user-images.githubusercontent.com/87318828/135167650-813f672d-37e0-49e1-8d01-1c4e5c988652.png">
?

**Comment by pixelzoom** [comment]
Absolutely, good catch @Nancy-Salpepi.

**Comment by pixelzoom** [comment]
Oh boy... here's another:

<img width="759" alt="screenshot_1297" src="https://user-images.githubusercontent.com/3046552/135168627-d2af8b24-18a0-4756-ba3e-d5c30ba00563.png">

**Comment by Nancy-Salpepi** [comment]
@pixelzoom I am noticing that keyboard nav doesn't allow me to tab to those amplitude boxes at all. Once I am in them, keyboard nav works, but shouldn't I be able to access them without a mouse?

**Comment by pixelzoom** [comment]
@Nancy-Salpepi I'm going to move that question to a new GitHub issue. It's a new topic.

**Comment by pixelzoom** [comment]
And another, weeeee! For this one, the button that opened it is then disabled, so it will be interesting to see where focus goes after closing it.

<img width="569" alt="screenshot_1298" src="https://user-images.githubusercontent.com/3046552/135169600-b6389615-5558-47ef-8ad9-50b7e91bf533.png">

**Comment by pixelzoom** [comment]
A couple of problems here.  Questions in _italics_ for @jessegreenberg:

(1) This dialog is opened by the sim when it detects a bad state:

<img width="759" alt="screenshot_1297" src="https://user-images.githubusercontent.com/3046552/135168627-d2af8b24-18a0-4756-ba3e-d5c30ba00563.png">

Adding `focusOnCloseButton` works fine. But when the dialog is closed, there's a focus problem - we don't know where to return focus to. Nothing in the UI gets focus, and pressing tab sets focus to the browser's URL textfield.  @jessegreenberg how should this be handled? 

(2) The RewardDialog has 2 button in addition to the close (X) button:

<img width="569" alt="screenshot_1298" src="https://user-images.githubusercontent.com/3046552/135169600-b6389615-5558-47ef-8ad9-50b7e91bf533.png">

There is currently no support for setting focus to those 2 buttons, so I've set it to the close button. It seems kind of odd to default to "close" when the dialog has content that you can interact with. _What is typical here - should focus start with some part of the dialog's interactive content?_

When you close the RewardDialog, it's also odd that the "Check Answer" button gets focus, because that button is not disabled.  _Is it OK to return focus to a disabled button?_

**Comment by jessegreenberg** [comment]
>  Nothing in the UI gets focus, and pressing tab sets focus to the browser's URL textfield.

This is expected. When the dialog is open, focus moves right to the close button even though highlights are still invisible. Pressing tab will then move focus to the next focusable element. In this case there is nothing available to focus so the browser moves focus to your URL textfield. I don't feel strongly but if you find it odd, feel free to make an issue in joist and we can potentially improve it.

> What is typical here - should focus start with some part of the dialog's interactive content?

Intuition tells me focus should be placed on the interactive content instead of the close button. In this case "New Level" makes sense to me. Master now has a `focusOnOpenNode` option and setter in Dialog, though that is not available in Fourier 1.0 SHAs. If this is something that needs to be in Fourier 1.0 you could focus the `newLevelButton` in the Dialog's `showCallback`. Or you could cherry pick changes from https://github.com/phetsims/sun/issues/719 once it is done.

**Comment by pixelzoom** [comment]
Thanks for the answers @jessegreenberg.

Re (1)... I'll do nothing, if this is expected behavior.

Re (2)... Since this involves modifying vegas.RewardNode either way, and that's common-code, it seems like I should do this correctly.  So I'll wait to cherry-pick phetsims/sun#719.  That should also allow me to remove all the calls to `focusCloseButton` that I added, since that is now the default - correct?

**Comment by pixelzoom** [comment]
Back to (1) - there's something really strange going on here.  

To demonstrate, step through these 2 scenarios, both of which open the same "Oops" dialog. Recall that this is a dialog that is opened by the sim, when it detects an invalid state.  That invalid state is an attempt to create a sawtooth waveform using cosines.

Scenario 1:
1. Start the sim, go to Discrete screen.
2. Set "Series" to "cos"
3. Tab to "Waveform" combo box, select "sawtooth"
4. The sim sets "Series" to "sin", then opens an "Oops" dialog. 
5. The dialog's close button has focus. Close the dialog using the keyboard, by pressing Esc or Space.
6. Note that there is now no focus. Pressing tab moves focus to the browser's URL textfield.

Scenario 2:
1. Start the sim, go to Discrete screen.
2. Set "Waveform" combo box to "sawtooth"
3. Tab to the "Series" radio button group and select "cos"
4. The sim sets "Series" to "sin", then opens an "Oops" dialog. 
4.  The dialog's close button has focus. Close the dialog using the keyboard, by pressing Esc or Space.
5. Note that focus is returned to "cos" button in the "Series" radio button group.

So depending on which component had focus when the dialog was opened, the sim may or may not have focus when the dialog is closed.  This inconsistency smells like unintentional or buggy behavior.  And it also seems like a usability problem - shouldn't focus be returned to the UI component that the user was interacting with, so they don't have to traversal all the way through the sim again?

@jessegreenberg please contact me on Zoom when you have a few minutes to discuss this.

**Comment by pixelzoom** [comment]
@arouinfar and I discuss https://github.com/phetsims/fourier-making-waves/issues/194#issuecomment-933685632 via Zoom. She agreed that the current behavior feels buggy.  Focus should be restored to the UI component that had focus at the time that the Oops dialog was opened.

**Comment by jessegreenberg** [comment]
Case 1 above is happening because the Node that had focus when the Dialog was opened is no longer focusable. When the Dialog opens the ComboBox closes and makes all of the ComboBoxListItemNodes invisible.

**Comment by pixelzoom** [comment]
@jessegreenberg and I talked about Case 1 in https://github.com/phetsims/fourier-making-waves/issues/194#issuecomment-933685632, and discovered that it's a problem in ComboBoxListBox.  The Property is modified before focus is restored to the ComboBoxButton.  That issues was addressed in phetsims/sun#721, and cherry-picked to Fourier 1.0 in the above commit.

**Comment by jessegreenberg** [comment]
In https://github.com/phetsims/sun/issues/719 we are removing `Dialog.focusCloseButton` and the last usages are the ones added to fourier for this issue. @pixelzoom can you please remove them and assign back to me when done so I can remove `focusCloseButton` from Dialog?

**Comment by pixelzoom** [comment]
@jessegreenberg I've removed calls to `focusCloseButton` in master and 1.0 branches in the above commit.

**Comment by pixelzoom** [comment]
RewardDialog has turned into a "keyboard navigation design" issue.  Because it's common code (vegas), I've created https://github.com/phetsims/vegas/issues/96.  This issue is on hold until that issue is addressed.

**Comment by pixelzoom** [comment]
Note to self.... Summary of what still needs to be done here:

* cherry-pick phetsims/sun#719  to Fourier 1.0
* verify that the default focus for dialogs is the close button
* complete keyboard navigation design and implementation for RewardDialog in phetsims/vegas#96
* cherry-pick vegas changes to Fourier 1.0

**Comment by pixelzoom** [comment]
@zepumph please let me know when https://github.com/phetsims/sun/issues/719 is closed, so that I can proceed with this issue.

**Comment by pixelzoom** [comment]
Cherry-picks identified in https://github.com/phetsims/sun/issues/719#issuecomment-937895180 have been completed and patched into Fourier 1.0 branch.   

This is ready for verification in the next RC.  This has been a long, complicated issue, so I'll need to summarize what and how to verify.

**Comment by Nancy-Salpepi** [comment]
Here is a video I took when testing (I only had 1 tab open to make it shorter):
![traversalorder](https://user-images.githubusercontent.com/87318828/136832689-78a54ddf-64e5-4577-8470-d6d670994886.gif)
